frame/support/src/storage/mod.rs

```diff
@@ -811,16 +815,292 @@ mod private {
 
 	impl<T: Encode> Sealed for Vec<T> {}
 	impl<Hash: Encode> Sealed for Digest<Hash> {}
+	impl<T: Value, S: Get<u32>> Sealed for BoundedVec<T, S> {}
 }
 
 impl<T: Encode> StorageAppend<T> for Vec<T> {}
 impl<T: Encode> StorageDecodeLength for Vec<T> {}
 
-/// We abuse the fact that SCALE does not put any marker into the encoding, i.e.
-/// we only encode the internal vec and we can append to this vec. We have a test that ensures
-/// that if the `Digest` format ever changes, we need to remove this here.
+/// We abuse the fact that SCALE does not put any marker into the encoding, i.e. we only encode the
+/// internal vec and we can append to this vec. We have a test that ensures that if the `Digest`
+/// format ever changes, we need to remove this here.
 impl<Hash: Encode> StorageAppend<DigestItem<Hash>> for Digest<Hash> {}
 
+/// Marker trait for types `T` that can be stored in storage as `Vec<T>`.
+pub trait Value: FullCodec + Clone + sp_std::fmt::Debug + Eq + PartialEq {}
+impl<T: FullCodec + Clone + sp_std::fmt::Debug + Eq + PartialEq> Value for T {}
+
+/// A bounded vector.
+///
+/// It implementations for efficient append and length decoding, as with a normal `Vec<_>`, once put
+/// into a [`StorageValue`].
+///
+/// As the name suggests, the length of the queue is always bounded. All internal operations ensure
+/// this bound is respected.
+#[derive(Encode, Decode, crate::DefaultNoBound, crate::CloneNoBound, crate::RuntimeDebugNoBound)]
+pub struct BoundedVec<T: Value, S: Get<u32>>(Vec<T>, sp_std::marker::PhantomData<S>);
```

Comment:
I guess that this is using a type number with `Get<u32>` instead of a const generic because we don't want to raise the min supported rust version to 1.51?

Reply:
Actually it probably can, I just didn't think about it :D Will try

> because we don't want to raise the min supported rust version to 1.51?

I am pretty sure we always require latest stable version and don't care much about older versions.

Reply:
errr -- no. 

I don't think we can make this configurable to the end-user, as it is now, with const generics. See how it is done in https://github.com/paritytech/substrate/pull/8580

Reply:
Right, ok. I think that in the future it may be worth a large-scale refactor to use actual consts and const generics instead of faking it with type parameters, but that may have to wait for Frame 3.0.

Reply:
We would already have switched to associated consts if it would be beneficial. The main advantage of using the trait is that we are much more flexible in tests. Overwriting constants would be a mess. Now we can just use thread local variables to overwrite the values. 

And than there are also test runtimes where the trait enables us to read variables from the storage and thus, make it really flexible. 

We would loose all this flexibility when using const generics.

Reply:
yeah, tests are also an issue with regard to const. Either of you know what kind of shenanigans we'd need to override consts? if possible at all?

Reply:
We would need to make tests more generic, add new traits that provide the consts we want to override and implement X times, everytime with the const value we want. You see where this is going. This would not be easy and would just require tons of boilerplate without any advantage.